### PR TITLE
Improve performance of analyzer driver in presence of symbol start an…

### PR DESCRIPTION
**Pull request by mavasani**
…alyzers. Primary changes:

1. Only execute `GetPerSymbolAnalyzerActionsAsync` for analyzers which have registered at least on symbol start action.
2. Only cache entries for _containing namespace or type symbols_ in `_perSymbolAnalyzerActionsMap` - we never access this map for member symbols.

## Performance numbers

Benchmark from https://github.com/dotnet/roslyn/blob/master/docs/wiki/Measuring-Compiler-Performance.md

1. Perf test baseline which has  **no SymbolStartAnalyzer**, using compiler built from **Roslyn master sources**
```
|                      Method |     Mean |    Error |   StdDev |   Median |      Min |      Max |       Gen 0 |      Gen 1 |     Gen 2 |  Allocated |
|---------------------------- |---------:|---------:|---------:|---------:|---------:|---------:|------------:|-----------:|----------:|-----------:|
| GetDiagnosticsWithAnalyzers | 14.091 s | 0.1028 s | 0.0961 s | 14.102 s | 13.927 s | 14.321 s | 228000.0000 | 62000.0000 | 2000.0000 | 1.36 GB |
```

2. Perf test baseline + **one trivial SymbolStartAnalyzer** (just registers a symbol start and end action), using compiler built from **Roslyn master sources**
```
|                      Method |     Mean |    Error |   StdDev |   Median |      Min |      Max |       Gen 0 |      Gen 1 |     Gen 2 |  Allocated |
|---------------------------- |---------:|---------:|---------:|---------:|---------:|---------:|------------:|-----------:|----------:|-----------:|
| GetDiagnosticsWithAnalyzers | 32.27 s  | 0.175 s  | 0.155 s  | 32.30 s  | 32.02 s  | 32.53 s  | 448000.0000 | 134000.0000| 4000.0000 |   2.57 GB  |
```

3. Perf test baseline + **one trivial SymbolStartAnalyzer**, using compiler built from the **branch in this PR**
```
|                      Method |    Mean |   Error |  StdDev |  Median |     Min |     Max |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|---------------------------- |--------:|--------:|--------:|--------:|--------:|--------:|------------:|-----------:|----------:|----------:|
| GetDiagnosticsWithAnalyzers | 15.77 s | 0.090 s | 0.080 s | 15.75 s | 15.63 s | 15.92 s | 236000.0000 | 68000.0000 | 2000.0000 |   1.38 GB |
```

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 1]
```diff
@@ -1831,24 +1839,37 @@ public bool HasSymbolStartedActions(AnalysisScope analysisScope)
...
-            // Compute additional inherited actions for this symbol by running the containing symbol's start actions.
-            AnalyzerActions inheritedActions = await getInheritedActionsAsync(this, symbol, analyzer, analysisStateOpt, cancellationToken).ConfigureAwait(false);
+            var allActions = await getAllActions(this, symbol, analyzer, analysisStateOpt, cancellationToken).ConfigureAwait(false);
+            return _perSymbolAnalyzerActionsCache.GetOrAdd((namespaceOrType, analyzer), allActions);
 
-            // Execute the symbol start actions for this symbol to compute additional actions for its members.
-            AnalyzerActions myActions = await getSymbolActionsCoreAsync(this, symbol, analyzer).ConfigureAwait(false);
-            AnalyzerActions allActions = !myActions.IsEmpty ? inheritedActions.Append(in myActions) : inheritedActions;
-            return _perSymbolAnalyzerActionsCache.GetOrAdd((symbol, analyzer), allActions);
+            static async ValueTask<AnalyzerActions> getAllActions(AnalyzerDriver driver, ISymbol symbol, DiagnosticAnalyzer analyzer, AnalysisState analysisStateOpt, CancellationToken cancellationToken)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]:
```suggestion
            static async ValueTask<AnalyzerActions> getAllActionsAsync(AnalyzerDriver driver, ISymbol symbol, DiagnosticAnalyzer analyzer, AnalysisState analysisStateOpt, CancellationToken cancellationToken)
```

**Comment by mavasani** [comment]
Ping @dotnet/roslyn-compiler for reviews.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

**Comment by mavasani** [comment]
Thank you @AlekseyTs @cston!

**Pull request merged**